experiment/oodatamodel/oodatamodel.go

```diff
@@ -0,0 +1,156 @@
+// Package oodatamodel contains the OONI data model.
+package oodatamodel
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"net"
+	"strconv"
+	"unicode/utf8"
+
+	"github.com/ooni/netx/x/porcelain"
+)
+
+// TCPConnectStatus contains the TCP connect status.
+type TCPConnectStatus struct {
+	Failure *string `json:"failure"`
+	Success bool    `json:"success"`
+}
+
+// TCPConnectEntry contains one of the entries that are part
+// of the "tcp_connect" key of a OONI report.
+type TCPConnectEntry struct {
+	IP     string           `json:"ip"`
+	Port   int              `json:"port"`
+	Status TCPConnectStatus `json:"status"`
+}
+
+// TCPConnectList is a list of TCPConnectEntry
+type TCPConnectList []TCPConnectEntry
+
+// NewTCPConnectList creates a new TCPConnectList
+func NewTCPConnectList(results porcelain.Results) TCPConnectList {
+	var out TCPConnectList
+	for _, connect := range results.Connects {
+		// We assume Go is passing us legit data structs
+		ip, sport, _ := net.SplitHostPort(connect.RemoteAddress)
+		iport, _ := strconv.Atoi(sport)
+		out = append(out, TCPConnectEntry{
+			IP:   ip,
+			Port: iport,
+			Status: TCPConnectStatus{
+				Failure: makeFailure(connect.Error),
+				Success: connect.Error == nil,
+			},
+		})
+	}
+	return out
+}
+
+func makeFailure(err error) (s *string) {
+	if err != nil {
+		serio := err.Error()
+		s = &serio
+	}
+	return
+}
+
+// HTTPTor contains Tor information
+type HTTPTor struct {
+	ExitIP   *string `json:"exit_ip"`
+	ExitName *string `json:"exit_name"`
+	IsTor    bool    `json:"is_tor"`
+}
+
+// HTTPBody is an HTTP body. We use this helper class to define a custom
+// JSON encoder that allows us to choose the proper representation depending
+// on whether the Value field is UTF-8 or not.
+type HTTPBody struct {
+	Value string
+}
+
+// MarshalJSON marshal the body to JSON following the OONI spec that says
+// that UTF-8 bodies are represened as string and non-UTF-8 bodies are
+// instead represented as `{"format":"base64","data":"..."}`.
+func (hb HTTPBody) MarshalJSON() ([]byte, error) {
```

Comment:
Note: you should also do this for binary header values.

Reply:
https://github.com/ooni/probe-engine/issues/110